src/main/java/org/opentripplanner/routing/impl/HighestFareInFreeTransferWindowFareServiceImpl.java

```diff
@@ -0,0 +1,82 @@
+package org.opentripplanner.routing.impl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Currency;
+import java.util.List;
+
+import org.opentripplanner.routing.core.Fare;
+import org.opentripplanner.routing.core.FareRuleSet;
+import org.opentripplanner.routing.core.Fare.FareType;
+import org.opentripplanner.routing.core.State;
+import org.opentripplanner.routing.edgetype.PatternInterlineDwell;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HighestFareInFreeTransferWindowFareServiceImpl extends DefaultFareServiceImpl {
+    private final boolean analyzeInterlinedTransfers;
+    private final int freeTransferWindowInMinutes;
+
+    public HighestFareInFreeTransferWindowFareServiceImpl(
+        Collection<FareRuleSet> regularFareRules, int freeTransferWindowInMinutes, boolean analyzeInterlinedTransfers
+    ) {
+        addFareRules(FareType.regular, regularFareRules);
+        this.freeTransferWindowInMinutes = freeTransferWindowInMinutes;
+        this.analyzeInterlinedTransfers = analyzeInterlinedTransfers;
+    }
+
+    private static final long serialVersionUID = 20120229L;
+    @SuppressWarnings("unused")
+    private static final Logger LOG = LoggerFactory.getLogger(HighestFareInFreeTransferWindowFareServiceImpl.class);
+
+    /**
+     * The fare calculation is designed to charge the rider incrementally as they use each service. The total cost of
+     * the itinerary will be equal to the leg of the journey that had the maximum fare. Additionally, a free transfer
+     * window is taken into account that will apply the current highest cost within the free transfer window and then
+     * reset to a new window and calculate additional free transfers from there.
+     */
+    @Override
+    protected boolean populateFare(Fare fare, Currency currency, FareType fareType, List<Ride> rides,
+                                   Collection<FareRuleSet> fareRules) {
+        float cost = 0;
+        float currentTransferWindowCost = 0;
+        long freeTransferWindowEndTime = -1;
+        for (Ride ride : rides) {
+            List<Ride> leg = new ArrayList<>();
+            leg.add(ride);
+            float rideCost = calculateCost(fareType, leg, fareRules);
+
+            if (ride.startTime > freeTransferWindowEndTime) {
+                // free transfer window has expired, add to overall cost
+                cost += currentTransferWindowCost;
+                // reset current window cost
+                currentTransferWindowCost = 0;
+                // reset transfer window end time to trigger recalculation in next block
+                freeTransferWindowEndTime = -1;
```

Comment:
In DefaultFareServiceImpl, the transfer window calculation seems to check the current ride's start time minus the first ride's start time (tripTime) against the transfer duration. However, this looks to check against the last ride and then resets the transfer window. Am I reading this correctly? My impression was that the free transfer window lasted from the "first tap" of a fare card (which would run counter to this resetting approach taken in this file), but maybe that's a mistaken impression.

https://github.com/ibi-group/OpenTripPlanner/blob/440701946004b85765303511d76a044242917f9a/src/main/java/org/opentripplanner/routing/impl/DefaultFareServiceImpl.java#L325-L328

Reply:
It's just a different calculation. Instead of comparing to the transfer duration, the `freeTransferWindowEndTime` is calculated as a point in time demarcating the end of the free transfer window. It is calculated by adding time onto the ride's start time as seen here:

https://github.com/ibi-group/OpenTripPlanner/blob/d6c5e934134768916c16aa1beb9ec51ce8d43983/src/main/java/org/opentripplanner/routing/impl/HighestFareInFreeTransferWindowFareServiceImpl.java#L58-L61

Reply:
Let's take an example. Free transfer window is one hour. Journey consists of three rides each take 45 minutes. Here's what I see when I read the code:

In the Default implementation, the trip time accumulates with each ride and is compared to duration:
1. ride 1 starts at trip time zero - transfer is N/A.
2. ride 2 starts at trip time 45 min - free transfer granted.
3. ride 3 starts at trip time 90 min - no transfer granted.

In the HighestFare... implementation, the transfer window is reset each time a ride is encountered:
1. ride 1 has `freeTransferWindowEndTime == -1` - transfer is N/A, window end set to T+60min.
2. ride 2 starts at trip time 45 min - free transfer granted, window end set to T+45+60min.
3. ride 3 starts at trip time 90 min - free transfer granted, window end set to T+45+45+60min.

Reply:
The window end time is only reset if the current ride's start time is later than the current window end time. Therefore, step 2 in your example is wrong because the window end time would not be updated.